### PR TITLE
Split punctuation semantic highlighting up into more tags

### LINKED ISSUE #6152: Add semantic token modifiers for punctuation
I used to have different colors for curly braces, operators, parens, brackets and other punctuation in TypeScript.

<details>
<summary>Take this colorization as an example:</summary>

image
</details>

It is not possible to achieve this amount of color diversity for punctuation in rust-analyzer for now, so I propose to add more specific semantic token modifiers for punctuation: `punctuation.curlyBrace`, `punctuation.bracket`, `punctuation.paren` ... so that it is possible to control the style of each individual punctuation symbol kind.

I am not insisting on implementing this via semantic token modifiers, this is just the first thing that comes to my head.

### PR DESCRIPTION
**Pull request by Veykril**
Open question would be the name of the delimiter modifiers. I chose them this was as I see them this way but from what I remember people tend to mix the names however they like. So maybe using `delimSquare`, `delimCurly`, `delimRound` would be better. That would also go well with `angle` becoming `delimAngle`?

Closes #6152

### CODE REVIEW

File: crates/ide/src/syntax_highlighting.rs [hunk 1]
```diff
@@ -599,7 +599,10 @@ fn highlight_element(
             _ if element.parent().and_then(ast::Attr::cast).is_some() => {
                 HighlightTag::Attribute.into()
             }
-            _ => HighlightTag::Punctuation.into(),
+            k => match punctuation_modifiers(k) {
+                Some(modifier) => Highlight::new(HighlightTag::Punctuation) | modifier,
```
Comment (matklad) [inline crates/ide/src/syntax_highlighting.rs:603]: Hm, I think we should have `Tag::Punctuation(Bracket | Brace ...)`, abusing modifiers for this *at this level* feels wrong.

But probably the best way to lower that to lsp in lsp_conv is indeed via modifiers: IIRC, they planned, but didn't ship inheritance for highlight tags. so we can't have `punctuation.brace`, where both things are tags. But better to double-check this.

**Comment by matklad** [comment]
Stats: blocked on moving kinds of punctuations to tags: https://github.com/rust-analyzer/rust-analyzer/pull/6238#discussion_r505522984

**Comment by matklad** [comment]
ping @Veykril )

**Comment by Veykril** [comment]
Kind of forgot about this PR 😅 This should be what you had in mind I believe right?

**Review by matklad — COMMENTED** [review]
bors r+

File: crates/ide/src/syntax_highlighting/tags.rs [hunk 2]
```diff
@@ -95,7 +117,17 @@ impl HlTag {
             HlTag::EscapeSequence => "escape_sequence",
             HlTag::FormatSpecifier => "format_specifier",
             HlTag::Keyword => "keyword",
-            HlTag::Punctuation => "punctuation",
+            HlTag::Punctuation(punct) => match punct {
+                HlPunct::Bracket => "bracket",
+                HlPunct::Brace => "brace",
+                HlPunct::Parenthesis => "parentheses",
```
Comment (lnicola) [inline crates/ide/src/syntax_highlighting/tags.rs:123]: Hmm.

This is the plural, I think.

Comment (Veykril) [reply crates/ide/src/syntax_highlighting/tags.rs:123]: I always mix those up

Comment (lnicola) [reply crates/ide/src/syntax_highlighting/tags.rs:123]: Oof, how do I run the HTML highlighting tests? Even with `RUN_SLOW_TEST=1 cargo test` nothing fails after changing that line.

Comment (lnicola) [reply crates/ide/src/syntax_highlighting/tags.rs:123]: Never mind, saving the changed files is a good idea :smile:.

**Pull request merged**